### PR TITLE
IWSLT2018 speech translation recipe

### PR DESCRIPTION
**Pull request by hirofumi0810**
### TODO
- [x] stage:0 Preprocessing
- [x] stage:1 Feature extraction
- [x] stage:2 dictionary & json
- [x] ASR & LM training & decoding (characater/BPE)
- [x] Speech translation & LM training & decoding (character/BPE)

### CODE REVIEW

File: egs/iwslt18st/st1/local/fix_trans.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env python2
+# -*- coding: utf-8 -*-
+
+# Copyright 2018 Nagoya University (Hirofumi Inaguma)
```
Comment (sw005320) [inline egs/iwslt18st/st1/local/fix_trans.py]: Kyoto University

File: egs/iwslt18st/st1/run.sh [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+# Copyright 2018 Kyoto University (Hirofumi Inaguma)
+#  Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)
+
+. ./path.sh
+. ./cmd.sh
+
+# general configuration
+backend=pytorch
+stage=0       # start from -1 if you need to start from data download
+gpu=            # will be deprecated, please use ngpu
```
Comment (sw005320) [inline egs/iwslt18st/st1/run.sh]: We removed these deprecated options in #335. Please follow them.

File: egs/iwslt18st/st1/run.sh [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+ngpu=0          # number of gpus ("0" uses cpu, otherwise use gpu)
+debugmode=1
+dumpdir=dump   # directory to dump full features
+N=0            # number of minibatches to be used (mainly for debugging). "0" uses all minibatches.
+verbose=0      # verbose option
+resume=        # Resume the training from snapshot
+
+# feature configuration
+do_delta=false # true when using CNN
```
Comment (sw005320) [inline egs/iwslt18st/st1/run.sh]: Remove the comment. Follow v.0.2.0 standard recipe.

File: egs/iwslt18st/st1/run.sh [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+feat_tr_dir=${dumpdir}/${train_set}/delta${do_delta}; mkdir -p ${feat_tr_dir}
+feat_dt_dir=${dumpdir}/${train_dev}/delta${do_delta}; mkdir -p ${feat_dt_dir}
+if [ ${stage} -le 1 ]; then
+    ### Task dependent. You have to design training and dev sets by yourself.
+    ### But you can utilize Kaldi recipes in most cases
+    echo "stage 1: Feature Generation"
+    fbankdir=fbank
+    # Generate the fbank features; by default 80-dimensional fbanks with pitch on each frame
+    for x in train dev2010 offlimit2018 tst2010 tst2011 tst2012 tst2013 tst2014 tst2015; do
+        steps/make_fbank_pitch.sh --cmd "$train_cmd" --nj 32 data/${x}_de exp/make_fbank/${x} ${fbankdir}
```
Comment (sw005320) [inline egs/iwslt18st/st1/run.sh]: Use utt2num_frames. Follow v.0.2.0 standard recipe.

File: egs/iwslt18st/st1/run.sh [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+         data/${train_dev}_trim ${dict} > ${feat_dt_dir}/data.json
+    for rtask in ${recog_set}; do
+        feat_recog_dir=${dumpdir}/${rtask}/delta${do_delta}
+        data2json.sh --feat ${feat_recog_dir}/feats.scp \
+        data/${rtask} ${dict} > ${feat_recog_dir}/data.json
+    done
+fi
+
+# You can skip this and remove --rnnlm option in the recognition (stage 3)
+lmexpdir=exp/train_rnnlm_2layer_bs256_de
```
Comment (sw005320) [inline egs/iwslt18st/st1/run.sh]: Add the backend variable. Follow v.0.2.0 standard recipe.

File: egs/iwslt18st/st1/run.sh [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+        --outdir ${lmexpdir} \
+        --train-label ${lmdatadir}/train.txt \
+        --valid-label ${lmdatadir}/valid.txt \
+        --epoch 60 \
+        --batchsize 256 \
+        --dict ${dict_en_char}
+fi
+
+if [ -z ${tag} ]; then
+    expdir=exp/${train_set}_${etype}_e${elayers}_subsample${subsample}_unit${eunits}_proj${eprojs}_d${dlayers}_unit${dunits}_${atype}_aconvc${aconv_chans}_aconvf${aconv_filts}_mtlalpha${mtlalpha}_${opt}_bs${batchsize}_mli${maxlen_in}_mlo${maxlen_out}
```
Comment (sw005320) [inline egs/iwslt18st/st1/run.sh]: Add backend variable. Follow v.0.2.0 standard recipe.

File: egs/iwslt18st/st1/run.sh [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+    if ${do_delta}; then
+        expdir=${expdir}_delta
+    fi
+else
+    expdir=exp/${train_set}_${tag}
```
Comment (sw005320) [inline egs/iwslt18st/st1/run.sh]: Add backend variable. Follow v.0.2.0 standard recipe.

File: egs/iwslt18st/st1/run_asr.sh [hunk 8]
```diff
@@ -0,0 +1,293 @@
+#!/bin/bash
```
Comment (sw005320) [inline egs/iwslt18st/st1/run_asr.sh]: Follow v.0.2.0 standard recipe.

File: egs/iwslt18st/st1/run.sh [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+etype=blstmp     # encoder architecture type
+elayers=6
+eunits=320
+eprojs=320
+subsample=1_2_2_1_1 # skip every n frame from input to nth layers
+# decoder related
+dlayers=1
+dunits=300
+# attention related
+atype=location
```
Comment (sw005320) [inline egs/iwslt18st/st1/run.sh]: Add adim option. See the example in the tedlium ASR recipe.

File: egs/iwslt18st/st1/run.sh [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+        ngpu=0
+
+        ${decode_cmd} JOB=1:${nj} ${expdir}/${decode_dir}/log/decode.JOB.log \
+            asr_recog.py \
+            --ngpu ${ngpu} \
+            --backend ${backend} \
+            --recog-json ${feat_recog_dir}/split${nj}utt/data.JOB.json \
+            --result-label ${expdir}/${decode_dir}/data.JOB.json \
+            --model ${expdir}/results/model.${recog_model}  \
+            --model-conf ${expdir}/results/model.conf  \
```
Comment (sw005320) [inline egs/iwslt18st/st1/run.sh]: `model.conf` -> `model.json`. Follow v.0.2.0 standard recipe.

**Comment by sw005320** [comment]
Great!
For the recipe in general, please follow the v.0.2.0 standard recipe. You could find the difference from v.0.1.3 by checking https://github.com/espnet/espnet/commits/dev/egs/an4/asr1

**Comment by sw005320** [comment]
Also send PR not to the master but the dev branch.

File: egs/iwslt18st/st1/run.sh (hunk 10 above)
Comment (sw005320) [reply egs/iwslt18st/st1/run.sh]: Corrections. Now this option is not required.

**Comment by hirofumi0810** [comment]
@sw005320,
BPE-based ST is almost done, but there are some bugs in transcripts.
I think this is due to disallowed whitespaces.

**Pull request merged**